### PR TITLE
DB token renewal before the token expires

### PR DESCRIPTION
**Pull request by chrise96**

### CODE REVIEW

File: val.py [hunk 1]
```diff
@@ -550,10 +552,6 @@ def run(
 
         callbacks.run('on_val_batch_end', batch_i, im, targets, paths, shapes, preds)
 
-    if skip_evaluation:
```
Comment (chrise96) [inline val.py:553]: This solves the issue that we had with the failed state!

File: database/database_handler.py [hunk 2]
```diff
@@ -111,3 +93,48 @@ def managed_session(self):
...
+            if sleep_time > 0:
+                LOGGER.info(f"Sleeping for {sleep_time} seconds until token renewal.")
+                time.sleep(sleep_time)
+
+            # Renew the token after the sleep
+            self._get_db_access_token()
+            LOGGER.info("Token for database renewed.")
+
+    def start_token_renewal_thread(self):
+        # TODO will this have conflicts with the threads used by YOLOv5?
```
Comment (chrise96) [inline database/database_handler.py:136]: Im not able to see the monitoring tab in Azure ML, I want to check the memory and CPU or GPU  usage

File: val.py [hunk 3]
```diff
@@ -264,7 +266,7 @@ def run(
 
             # Perform database operations using the 'session'
             # The session will be automatically closed at the end of this block
-            with db_config.get_session() as session:
+            with db_config.managed_session() as session:
```
Comment (chrise96) [inline val.py:269]: This I forgot to change. We use managed_session now thats a wrapper over get_session

File: database/database_handler.py [hunk 4]
```diff
@@ -21,26 +21,31 @@ def __init__(self, db_username, db_hostname, db_name):
         self.db_username = db_username
         self.db_hostname = db_hostname
         self.db_name = db_name
+        self.access_token = None
+        self.token_expiration_time = None
+        self.token_renewal_margin = timedelta(minutes=5)  # Set the token renewal margin
```
Comment (sdavrieux) [inline database/database_handler.py:26]: Comment unnecessary since the naming is good 😉

File: database/database_handler.py [hunk 5]
```diff
@@ -111,3 +93,48 @@ def managed_session(self):
             raise e
         finally:
             session.close()
+
+    def close_connection(self):
+        try:
+            # Dispose the engine
```
Comment (sdavrieux) [inline database/database_handler.py:99]: Redundant comment

File: database/database_handler.py [hunk 6]
```diff
@@ -111,3 +93,48 @@ def managed_session(self):
...
+            self.engine.dispose()
```
Comment (sdavrieux) [inline database/database_handler.py:100]: Do we need to manually dispose the engine?

File: database/database_handler.py [hunk 7]
```diff
@@ -111,3 +93,48 @@ def managed_session(self):
...
+        except SQLAlchemyError as e:
+            LOGGER.info(f"Error disposing the database engine: {str(e)}")
+            raise e
+
+    @staticmethod
+    def extract_upload_date(path):
```
Comment (sdavrieux) [inline database/database_handler.py:106]: This doesn't seem the right place to this function. Database handler should handle database connection and queries. This is more project related. Especially if in the future we want to extract the database handler to its own repo.

File: database/database_handler.py [hunk 8]
```diff
@@ -111,3 +93,48 @@ def managed_session(self):
...
+            image_upload_date = datetime.strptime(date_time_str, "%Y-%m-%d_%H:%M:%S")
+            image_upload_date = image_upload_date.strftime("%Y-%m-%d")
+        except ValueError as e:
+            LOGGER.info(f"Invalid folder structure, can not retrieve date: {date_time_str}")
+            raise e
+
+        return image_filename, image_upload_date
+
+    def _renew_token_periodically(self):
+        while True:
```
Comment (sdavrieux) [inline database/database_handler.py:121]: A bit scared of this way I have to say. 
Cannot we just do the check when we use the database handler?
For example when querying we check if the token is expired and then we query.

File: database/database_handler.py [hunk 9]
```diff
@@ -111,3 +93,48 @@ def managed_session(self):
...
+                time.sleep(sleep_time)
+
+            # Renew the token after the sleep
+            self._get_db_access_token()
+            LOGGER.info("Token for database renewed.")
+
+    def start_token_renewal_thread(self):
+        # TODO will this have conflicts with the threads used by YOLOv5?
+        # Start a thread to renew the token periodically
+        token_renewal_thread = Thread(target=self._renew_token_periodically)
```
Comment (sdavrieux) [inline database/database_handler.py:138]: Threading can be a pain, check my other comment.

File: val.py [hunk 10]
```diff
@@ -264,7 +266,7 @@ def run(
...
                 try:
                     # Construct the query to get all rows with a certain processing status
                     query = session.query(
```
Comment (sdavrieux) [inline val.py:272]: Here for example: check the token validity, renew it if needed and then query.

File: database/database_handler.py (hunk 6 above)
Comment (chrise96) [reply database/database_handler.py:99]: The pipeline failed when I used close_connection

The "with statement" in val.py (with db_config.managed_session() as session) ensures that the session is properly closed and any resources associated with it are released when you exit the block. SQLAlchemy takes care of closing the session, so we don't need to manually call close or dispose.

Remove this function?

File: database/database_handler.py (hunk 7 above)
Comment (chrise96) [reply database/database_handler.py:105]: Move it to val.py? Or make a new "parse.py" class in the database folder?

File: database/database_handler.py (hunk 8 above)
Comment (chrise96) [reply database/database_handler.py:120]: We do a lot of querying, checken each time the things we do in _renew_token_periodically will give a lot of overhead. I will think about other options

File: val.py [hunk 11]
```diff
@@ -303,6 +305,8 @@ def run(
                                                        prefix=colorstr(f'{task}: '))
 
         if skip_evaluation:
+            db_config.validate_token_status() # TODO only in the inference for loop
```
Comment (chrise96) [inline val.py:308]: @sdavrieux

File: val.py [hunk 12]
```diff
@@ -455,6 +459,8 @@ def run(
                         im_orig[si][y1:y2, x1:x2] = blurred
 
                         if skip_evaluation:
+                            db_config.validate_token_status() # TODO is this the right location? or line 453
```
Comment (chrise96) [inline val.py:462]: @sdavrieux

File: yolov5/val.py [hunk 13]
```diff
@@ -455,11 +459,12 @@ def run(
                         im_orig[si][y1:y2, x1:x2] = blurred
 
                         if skip_evaluation:
+                            db_config.validate_token_status() # TODO is this the right location? or line 453
```
Comment (epureanudiana) [inline yolov5/val.py:462]: should we remove the comment?

**Review by sdavrieux — APPROVED** [review]

**Pull request merged**